lib/decoder.js

```diff
@@ -33,29 +38,60 @@ async function readVarint (reader) {
 
 /**
  * @param {BytesReader} reader
- * @returns {Promise<CarHeader>}
+ * @returns {Promise<CarV2FixedHeader>}
  */
-export async function readHeader (reader) {
+async function readV2Header (reader) {
+  /** @type {Uint8Array} */
+  const bytes = await reader.exactly(V2_HEADER_LENGTH)
+  const dv = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength)
+  let offset = 0
+  const header = {
+    version: 2,
+    /** @type {[bigint, bigint]} */
+    characteristics: [
+      dv.getBigUint64(offset, true),
+      dv.getBigUint64(offset += 8, true)
+    ],
+    dataOffset: Number(dv.getBigUint64(offset += 8, true)),
+    dataSize: Number(dv.getBigUint64(offset += 8, true)),
+    indexOffset: Number(dv.getBigUint64(offset += 8, true))
+  }
+  reader.seek(V2_HEADER_LENGTH)
+  return header
+  /* c8 ignore next 2 */
+  // Node.js 12 c8 bug
+}
+
+/**
+ * @param {BytesReader} reader
+ * @param {number} [strictVersion]
+ * @returns {Promise<CarHeader|CarV2Header>}
+ */
+export async function readHeader (reader, strictVersion) {
   const length = await readVarint(reader)
   const header = await reader.exactly(length)
   reader.seek(length)
   const block = decodeDagCbor(header)
-  if (block == null || Array.isArray(block) || typeof block !== 'object') {
+  if (!headerValidator(block)) {
     throw new Error('Invalid CAR header format')
   }
-  if (block.version !== 1) {
-    if (typeof block.version === 'string') {
-      throw new Error(`Invalid CAR version: "${block.version}"`)
-    }
+  if ((block.version !== 1 && block.version !== 2) || (strictVersion !== undefined && block.version !== strictVersion)) {
     throw new Error(`Invalid CAR version: ${block.version}`)
   }
-  if (!Array.isArray(block.roots)) {
+  // we've made 'roots' optional in the schema so we can do the version check
+  // before rejecting the block as invalid if there is no version
+  const hasRoots = Array.isArray(block.roots)
+  if ((block.version === 1 && !hasRoots) || (block.version === 2 && hasRoots)) {
     throw new Error('Invalid CAR header format')
   }
-  if (Object.keys(block).filter((p) => p !== 'roots' && p !== 'version').length) {
-    throw new Error('Invalid CAR header format')
+  if (block.version === 1) {
+    return block
   }
-  return block
+  // version 2
+  const v2Header = await readV2Header(reader)
+  reader.seek(reader.pos - v2Header.dataOffset)
+  const v1Header = await readHeader(reader, 1)
```

Comment:
could this lead to recursive weirdness where you put a v2 in a v2?

Reply:
shouldn't be, that's what the optional second argument is for in this case - the `block.version !== strictVersion` should mean that if it's not a v1 header then it'll error before it even gets down here on the recursive call
not a bad idea for a test though

Reply:
added test for failing a recursive carv2 @ ad8cdc0
good point to make explicit in the spec I think